[MRG+1] Reproducible PS/PDF output (master)

JojoBoulix opened the pull request:
This is a rebase of #6595 to the master branch.

Several software packages use matplotlib in their building process (mainly to produce PS or PDF documents). To make their build reproducible, it would be great to make matplotlib output reproducible.

To allow reproducible PS and PDF output:
- honour SOURCE_DATE_EPOCH for timestamps in PS and PDF files.
  See https://reproducible-builds.org/specs/source-date-epoch/
- get keys sorted so that hatch patterns, images and markers are included with
  a reproducible order in the PDF file. Another solution is to sort `self.hatchPatterns` in `writeHatches` (and similar ordering in `writeImages` and `writeMarkers`), but this consumes more memory. 

This patch has been submitted in debian bug [#827361](https://bugs.debian.org/827361)

See also https://reproducible-builds.org/

tacaswell commented:
I think we are already sorting all of those keys on the way out so do not need the ordered dicts.

Can you add an entry in https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new and a test (like the deterministic svg test) that this actually works?

attn @jkseppan

jkseppan commented:
I agree that a test is needed. I don't see the reason for the ordered dicts either, but there could be some subtlety that I'm missing. It would be neat to make the date settable by the user (via e.g. the PdfPages constructor) but of course not necessary for this particular use case.

JojoBoulix commented:
I added a test for reproducible PDF output (you're right: this is very important!).
The test shows that the ordered dicts are needed for markers and hatches, due to the dict iterations at the beginning of `writeHatches` and `writeMarkers`, which are unreproducible with python3.
Even if the same iteration occurs in `writeImages`, the tests passes without ordering `_images`. I think this is because the `_images` keys are integers.
I also added two orderings (related to fonts), that appeared to be essential thanks to the test.

jkseppan commented:
Oh, of course! I was thinking about the page stream, but it also matters in which order the objects are output at the top level of the pdf file. I suspect that the ordering is needed for `_images` too - maybe the dict implementation is a list for integer-keyed small dictionaries, or something.

JojoBoulix commented:
I would like to add tests for postscript too, but I'm afraid the code will be nearly the same as for PDF: do you think I should move the common code somewhere (maybe in a new `testing/reproducible.py` file), or duplicate it in `test_backend_ps.py` and `test_backend_pdf.py`?

tacaswell commented:
```
.SK...
======================================================================
FAIL: Test SOURCE_DATE_EPOCH support for PS output
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Miniconda-x64\envs\test-environment\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "c:\projects\matplotlib\lib\matplotlib\testing\decorators.py", line 152, in wrapped_callable
    func(*args, **kwargs)
  File "c:\projects\matplotlib\lib\matplotlib\tests\test_backend_ps.py", line 186, in test_source_date_epoch
    _test_source_date_epoch("ps", b"%%CreationDate: Sat Jan  1 00:00:00 2000")
  File "c:\projects\matplotlib\lib\matplotlib\testing\determinism.py", line 122, in _test_source_date_epoch
    assert string in buff
AssertionError
--------------------------------------------------------------------
```

These failures on appveyor look real.

tacaswell commented:
My naive guess is that there is a unicode vs bytes issues

JojoBoulix commented:
> These failures on appveyor look real.

Sure! I'm working on it but for now I can't reproduce the problem at home.

JojoBoulix commented:
I think this is better now.
* For parallel tests, I think threads are used, and they share the same environment. So setting the SOURCE_DATE_EPOCH environment variable in a test is not good. So I used subprocesses instead.
* The timestamp was 'Jan  1' or 'Jan 01' sometimes, so I used Dec 15 instead to solve this issue.

Appveyor is now reporting only a `HTTPError: 500 Server Error`. I think this is not related to Matplotlib code, but I don't know if I can restart the build.

tacaswell commented:
:+1: from me, attn @jkseppan can you do a final review?

jkseppan commented on doc/users/whats_new/reproducible_ps_pdf.rst:
```diff
@@ -0,0 +1,6 @@
+
+The ``SOURCE_DATE_EPOCH`` environment variable can now be used to set
+the timestamps value in the PS and PDF outputs, which are then
+reproducible.
```
We should list the known limitations, because when downstream users see this promise, they may start depending on it and run into the limitations later. At least usetex with the ps backend should be ruled out, based on comments in the tests.

jkseppan replied:
Please add a link to the specification that describes the value of this environment variable.

jkseppan replied:
Other possible limitations include at least mathtext, ps.usedistiller, ps.fonttype, and pdf.fonttype. These are features or settings that I think could easily hide a source of nondeterminism and that don't seem to get exercised by the new test. (To get a more reliable list of relevant features than just my guesses, you'd need to run the test with coverage checking and see which parts of the backends get no coverage.)

To be clear, I don't mean we should delay merging until there are tests for everything, but we should be realistic in how we document this. I imagine the reproducible-builds community would appreciate the feature as it is now, with careful documentation of what exactly they can rely on.

jkseppan commented:
Commit c007f4998a3e4ab411c8dd58ab186ced4c179ca7 changes the tests to use a two-digit date in December instead of January 1st. I think the underlying problem should be fixed instead of avoiding it in the tests, since we don't want to tell users that their `SOURCE_DATE_EPOCH` values must be two-digit days.

JojoBoulix commented:
I considered reproducibility this way: with the same tools (same versions of all what you installed on your computer), the same commands lead to the same results.
With this point of view, I accept that with some different versions/implementations of python, the date can be written different ways (leading space or 0 for example). The problem with the tests was that we face different implementations with different behaviors. We don't pretend that the output will be exactly the same with all future versions of all the tools, all implementations of python.
I also accept that ghostscript can change its way to format the timestamp in a future version, but I did not include a test with usetex relying to the current behavior of ghostscript. Maybe this is too prudent.
Let me try to be more explicit about that in the doc.

jkseppan commented on doc/users/whats_new/reproducible_ps_pdf.rst:
```diff
@@ -2,5 +2,12 @@ Reproducible PS and PDF output
+Matplotlib does its best to make PS and PDF outputs reproducible, but
+be aware that some unreproducibility issues can arise if you use
+different versions of Matplotlib and the tools it relies on. Although
+standard plots has been checked to be reproducible, external tools can
```
This is sort of nitpicking, but mathtext and the fonttype settings are internal to matplotlib. Here's my suggestion:

The reproducibility of the output from the PS and PDF backends has so far been tested using various plot elements but only default values of options such as `{ps,pdf}.fonttype`  that can affect the output at a low level, and not with the mathtext or usetex features. When matplotlib calls external tools (such as PS distillers or LaTeX) their versions need to be kept constant for reproducibility, and they may add sources of nondeterminism outside the control of matplotlib.

JojoBoulix replied:
Thanks for the suggestion!

QuLogic commented:
Instead of `time`, you could use `datetime`, and then specify an explicit formatting string to avoid differences, no?

JojoBoulix commented:
> Instead of time, you could use datetime, and then specify an explicit formatting

You're right.

jkseppan commented:
:+1: thanks @JojoBoulix! I think this is ready for merging.

JojoBoulix commented:
I think I have to get rid of the AppVeyor issues first… I'm afraid I'm a little stuck on this.

JojoBoulix commented:
Oh yes I see:
```
  int(source_date_epoch) ).strftime("%a %b %e %T %Y")
ValueError: Invalid format string
```
In some environments one of these formats is not available (maybe `%e`?)…

dopplershift commented:
I don't see `e`, `T`, or `Y` listed here: https://docs.python.org/3.5/library/datetime.html#strftime-and-strptime-behavior

JojoBoulix commented:
`Y` is, but not `e` nor `T`: thanks. So I think it is better to use the zero-padded `%d`, which is standard.

Kojoley commented:
Can we have a `rcParams` variable for this? Like one that `svg` have (I am talking about `svg.hashsalt`)

Kojoley commented on lib/matplotlib/tests/test_backend_pdf.py:
```diff
@@ -12,6 +12,8 @@
 from matplotlib import cm, rcParams
 from matplotlib.backends.backend_pdf import PdfPages
 from matplotlib import pyplot as plt
+from matplotlib.testing.determinism import (_test_source_date_epoch,
```
I do not think it is a good idea to split tests into multiple files (apparently this will be the first in the whole matplotlib)

JojoBoulix replied:
There is common code in PDF and PS backends tests, that I wrote in `determinism.py`. Where do you think it should be?
I renamed the functions to remove the `test' keyword

Kojoley commented on lib/matplotlib/testing/determinism.py:
```diff
@@ -0,0 +1,136 @@
+    import sys
+    from subprocess import check_call
+    filename = 'test_SDE_on.%s' % format
+    check_call([sys.executable, '-R', '-c',
```
Why this cannot be done without subprocesses?

Kojoley replied:
I think that subprocess call can be replaced by function with `switch_backend` decorator

QuLogic replied:
I think the point is to use different random number generator seeds, not to segregate the backend changes.

Kojoley commented on lib/matplotlib/testing/determinism.py:
```diff
@@ -0,0 +1,136 @@
+    x = range(5)
+    fig.add_subplot(1, 6, 6).plot(x, x)
+
+    fig.savefig(filename, format=format)
```
Is not it better to save content into `BytesIO`? Then you can simply compare two variables (no more need of the temporary files).

JojoBoulix replied:
I tried but did not succeed to make it work without using a temporary file.

QuLogic replied:
Since they're running in external processes, I don't think `BytesIO` would be the right approach. What might work is to output to `stdout` and read from there with `check_output`. I can open a PR to show how this might work.

Kojoley replied:
> Since they're running in external processes

At the time the comment was written I have tried to use `multiprocessing` and thought it should work #6927, but not for Linux (I just have used an implementation defined behavior in Windows  `multiprocessing`)

> What might work is to output to `stdout` and read from there with `check_output`. I can open a PR to show how this might work.

It is a good idea, I like it.

NelleV commented:
The tests are failing on this PR.

QuLogic commented:
`nose` seems to be incomplete in all those failures, which seems odd. Also, is there a reason we're using a forked nose for testing?

Kojoley commented:
> `nose` seems to be incomplete in all those failures, which seems odd. 

It is relative import problem. `lib/matplotlib/testing/determinism.py` does not include `from __future__ import absolute_import`.

> Also, is there a reason we're using a forked nose for testing?

AFAIK the fork contains Python 3.5 and coverage support fixes.

Kojoley commented on lib/matplotlib/testing/determinism.py:
```diff
@@ -53,7 +55,13 @@ def _determinism_save(objects='mhi', format="pdf"):
+        stdout = io.StringIO()
+    else:
+        stdout = getattr(sys.stdout, 'buffer', sys.stdout)
+    fig.savefig(stdout, format=format)
```
Why not just `fig.savefig(getattr(sys.stdout, 'buffer', sys.stdout), format=format)`?

QuLogic commented:
Ah yes, that file is missing all `__future__` imports.

JojoBoulix commented:
After some little corrections, it seems to me that failed tests reported by travis and appveyor are not related to the pull request. However I don't understand how they appeared…

QuLogic commented:
Both AppVeyor fails are fixed on `master`; I restarted Travis. A rebase or new commit should update that.

JojoBoulix commented:
The first commit af4213ed07f1adaa77c9aaa8160904437ad8feeb is a bugfix: the `text.usetex` setting was not properly passed to `_determinism_save` before.
With the commit bf7387e95fb26615e0a7bfeb80f7fdc0c9404318, the failed tests with `text.usetex` on should be skipped, because I think the failure is due to `ghostscript` (some versions now honors `SOURCE_DATE_EPOCH`, but some don't). I still kept the test because it can be interesting to know if the test passes or not.

Kojoley commented:
> I think the failure is due to `ghostscript` (some versions now honors `SOURCE_DATE_EPOCH`, but some don't).

The fail happens on the build where ghostscript is not installed, so bf7387e95fb26615e0a7bfeb80f7fdc0c9404318 commit is not the right way to handle such situation.

JojoBoulix commented:
The `[XPASS(strict)] This test needs a ghostscript installation` fail happened because the test should have failed (tagged `@needs_ghostscript`) but did not fail, because `usetex` was not properly passed to `_determinism_save` before af4213ed07f1adaa77c9aaa8160904437ad8feeb, so that the test ran without using ghostscript.
Correcting this, I then saw that now that `usetex` is properly passed, the test using `usetex` failed because of `ghostscript`, which does not honour `SOURCE_DATE_EPOCH` in the installed version. So I applied bf7387e95fb26615e0a7bfeb80f7fdc0c9404318 so that we can see if the test pass (good `ghostscript` version) or not, but without failing.

tacaswell commented:
'power cycled to restart CI (appveyor failures looked due to qt-related packaging issues).

tacaswell commented on lib/matplotlib/backends/backend_pdf.py:
```diff
@@ -135,6 +136,20 @@ def _string_escape(match):
 
 
+# tzinfo class for UTC
+class UTCtimezone(tzinfo):
```
We already have this in mpl/dates.py

tacaswell replied:
https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/dates.py#L155

JojoBoulix replied:
Yes! Sorry I missed that.

tacaswell commented:
Other than one minor change, this looks good to go!

Don't worry about the coverall failure, that is un-related.

Sorry this has dragged on so long.

QuLogic commented:
The coveralls drop is probably due to a change in master, and not this PR; a rebase to latest master might help that, but is not strictly necessary.

QuLogic commented on doc/users/whats_new/reproducible_ps_pdf.rst:
```diff
@@ -0,0 +1,14 @@
+------------------------------
+
+The ``SOURCE_DATE_EPOCH`` environment variable can now be used to set
+the timestamps value in the PS and PDF outputs. See
```
timestamp

QuLogic commented on lib/matplotlib/testing/determinism.py:
```diff
@@ -0,0 +1,147 @@
+        default value is "mhi", so that the test includes all these objects.
+    format : str
+        format string. The default value is "pdf".
+    uid : str
```
Not seeing this being used for anything?

JojoBoulix replied:
The `uid` option is used in `test_determinism_all_tex` from `lib/matplotlib/tests/test_backend_ps.py`

JojoBoulix replied:
But it was only useful when files were used to store the figures. I will remove that.

QuLogic commented on lib/matplotlib/testing/determinism.py:
```diff
@@ -0,0 +1,147 @@
+        format string, such as "pdf".
+    string : str
+        timestamp string for 2000-01-01 00:00 UTC.
+    keyword : str
```
It appears to be `bytes`, not `str`.

JojoBoulix commented:
@QuLogic: thanks a lot for your review. I corrected those points.

QuLogic commented:
A rebase might fix coveralls (as I'm pretty sure it was a change in master that dropped coverage), but it's not strictly needed. I guess we aren't really waiting on anything else here.

JojoBoulix commented:
Thanks to you all for your kind support!

tacaswell commented:
@JojoBoulix Thanks for taking care of this!  Sorry it was such a protracted process.
